src/PdoConnection.php

```diff
@@ -58,6 +58,16 @@ class PdoConnection
     private $reuseConnection = true;
 
     /**
+     * @var int Number of open transactions
+     */
+    private $transactionCount = 0;
+
+    /**
+     * @var bool Flag that tells whether a rollback is needed
+     */
+    private $rollbackNeeded = false;
```

Comment:
$transactionCount represents how many transactions are currently open.  $rollbackNeeded is only true when a rollback is called on a transaction within another transaction, which is very different from $transactionCount === 0.  I means that if the outer transaction tries to commit, we need to rollback instead.